enhancements/ingress/configurable-dns-loadbalancerservice.md

```diff
@@ -0,0 +1,343 @@
+the cluster admin to manually delete it. 
+
+The support provided by the installer and this new field aren't entirely
+mutually exclusive, i.e. if customer has opted to disable DNS management cluster
+wide (as documented
+[here](https://github.com/openshift/installer/blob/master/docs/user/aws/install_upi.md#remove-dns-zones)),
+this field is of no value and need not be set since *DNSRecord* created is silently
+reconciled by the DNS controller.
+
+__Note:__ Appropriate logging is a must on all controllers clearly indicating the
+current DNS management policy set on both ingresscontroller and DNSRecord and any
+subsquent effects such as skipping reconciliation, etc must also be logged.
+
+### Risks and Mitigations
+
+This feature does not support default ingresscontrollers, if updated to be unmanaged
+in conjunction with updating the domain could result in breaking cluster connectivity
+if not all components are properly updated. This should be done at the discretion
+of the cluster admin. The manual creation of the required DNS records will need to be
+done prior to making this change, as the canary controller will try to ensure that the
+ingress domains are reachable, and otherwise this sets the `cluster-ingress-operator`
+in a degraded state.
+
+### Drawbacks
+
+After an ingresscontroller is updated from `Managed` to `Unmanaged`, the previous
+*DNSRecord* is left in an `Unmanaged` state which could result in orphaned resources
+on the cluster and the cloud provider.
+
+## Design Details
+
+### Open Questions
+
+Currently, as per documentation `.spec.domain` in the ingresscontroller cannot be
+updated after creation since updating the [domain](https://github.com/openshift/cluster-ingress-operator/blob/972f09b9dbb181ae5c414da2a990b57c60fde9d8/pkg/operator/controller/ingress/controller.go#L342-L355)
+in the ingresscontroller doesn't result in [updating](https://github.com/openshift/cluster-ingress-operator/blob/972f09b9dbb181ae5c414da2a990b57c60fde9d8/pkg/operator/controller/ingress/dns.go#L97)
+the wildcard *DNSRecord*. 
+> 1. Can this logic be changed to support moving from managed DNS to
+     unmanaged DNS on the `default` ingresscontroller?
+     
+  __Closed:__ The operator will continue to retain immutability of `.spec.domain`
+              and will not support updating domain when migrating from `Managed` to 
+              `Unmanaged` DNS.
+  
+
+### Test Plan
+
+- Test by creating a new secondary ingresscontroller with a custom domain
+  and `.loadbalancer.dnsManagementPolicy` set to `Unmanaged`. Ensure proper conditions
+  and connectivity.
+- Test deletion of ingresscontrollers to ensure correct behaviour of the associated
+  *DNSRecord* CR and the DNS record on the cloud provider.  
+- Test the following update paths on a custom ingresscontroller
+  - Updating `.loadbalancer.dnsManagementPolicy` from `Managed` -> `Unmanaged` -> `Managed` and
+    to ensure no conflicts during creation/recreation of DNS record on the cloud provider.
```

Comment:
Why "recreation"?  Is the test going to delete the DNS record after setting `dnsManagementPolicy` to "Unmanaged"?

Reply:
Yes, that is what I had in mind. Thinking of testing both the scenarios to verify there were no collisions during recreation.

Reply:
Thanks, that makes sense; it just isn't clear from the test plan as written.

Reply:
The test plan might need to be revisited. I've mostly implemented this feature and turns out we can't have updates from `Unmanaged` to `Managed` since `domain` cannot be updated. Essentially, to update an ingresscontroller from `Managed` to `Unmanaged` is a DELETE and CREATE. 

Although it is possible to update from `Managed` to `Unmanaged` users would be forced to re-use the domain leading to again DELETE and CREATE of the ingresscontroller so that a new unmanaged domain can be given. 

So I'm thinking it is better to make `dnsManagementPolicy` also not updatable similar to `domain` (figured this would be better than making `domain` updatable`), then the test plan would be straight forward. Just test with an Unmanaged ingresscontroller.

Reply:
I don't understand.  Why would it matter that `domain` cannot be updated?  The operator should still be able to do an upsert when the management state is changed from `Unmanaged` to `Managed`, the same as it does an upsert if a managed DNSRecord's targets change.  On some platforms, an upsert may actually be implemented as discrete delete and create steps within a transaction, but that's an implementation detail that is abstracted behind the `dns.Provider` interface.